File: day:25/sortedArrayToBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function(nums) {
    // base cases
    if(nums.length === 0) return null;
    if(nums.length === 1) {
        return new TreeNode(nums[0]);
    }
    // find the mid of the array
    let mid = Math.floor(nums.length/2);
    let left = nums.slice(0, mid);
    let right = nums.slice(mid + 1);
    
    let node = new TreeNode(nums[mid]);
    node.left = sortedArrayToBST(left);
    node.right = sortedArrayToBST(right);
    return node;
};

File: day:25/sortedArrayToBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        // base cases
        if(nums.length == 0) return null;
        if(nums.length == 1) {
            return new TreeNode(nums[0]);
        }
        // find the mid of the array
        var mid = (int)Math.floor(nums.length/2);
        var left = Arrays.copyOfRange(nums, 0, mid);
        var right = Arrays.copyOfRange(nums, mid + 1, nums.length);
    
        var node = new TreeNode(nums[mid]);
        node.left = sortedArrayToBST(left);
        node.right = sortedArrayToBST(right);
        return node;
        
    }
}
